presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/JoinPlannerUtils.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.iterative.rule;
+
+import com.facebook.presto.sql.planner.iterative.Lookup;
+import com.facebook.presto.sql.planner.optimizations.PlanNodeSearcher;
+import com.facebook.presto.sql.planner.plan.AggregationNode;
+import com.facebook.presto.sql.planner.plan.AssignUniqueId;
+import com.facebook.presto.sql.planner.plan.EnforceSingleRowNode;
+import com.facebook.presto.sql.planner.plan.JoinNode;
+import com.facebook.presto.sql.planner.plan.LimitNode;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.ProjectNode;
+import com.facebook.presto.sql.planner.plan.TableScanNode;
+import com.facebook.presto.sql.planner.plan.ValuesNode;
+import com.facebook.presto.util.MorePredicates;
+
+final class JoinPlannerUtils
+{
+    private JoinPlannerUtils() {}
+
+    /**
+     * Returns true if join build side can be replicated. This depends
+     * if build side consists of nodes for which stats shouldn't be
+     * underestimated.
+     */
+    public static boolean canReplicate(JoinNode joinNode, Lookup lookup)
+    {
+        return canReplicate(joinNode.getRight(), lookup);
+    }
+
+    public static boolean canReplicate(PlanNode joinBuildSide, Lookup lookup)
+    {
+        return PlanNodeSearcher.searchFrom(joinBuildSide, lookup)
+                .where(MorePredicates.isInstanceOfAny(TableScanNode.class, ValuesNode.class))
+                .recurseOnlyWhen(MorePredicates.isInstanceOfAny(
```

Comment:
I was trying to put my thoughts together for a while, will try to explain what do i think.

Currently i can see only a singe rule that tends to underestimate in some cases. This is `FilterStatsRule` (in fact `FilterStatsCalculator`). Also because `JoinStatsRule` relies on `FilterStatsCalculator`, the `JoinStatsRule` may also underestimate.

From what i understand, most of the underestimation in `FilterStatsCalculator` are because of the heuristic we use for inequality predicate estimates, as it is very hard to correctly estimate inequality predicates without histograms:

https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/cost/StatisticRange.java#L31

I agree that equality predicates can also underestimate in case of non uniformly distributed data, but it is less likely.

Currently i'm actively working on improving `FilterStatsCalculator`. Especially i'm trying to get rid of wild guesses when there is no driving equality predicate. It should improve situation with miss-estimations.

For the nasty cases when we are still underestimating, the user can just disable CBO, and set the `BROADCAST` distribution manually. Given that our model is not ideal, that shouldn't be very unusual situation.

I can the value of this change. But from the other side that's another user visible knob with a very limited usability and an area of application. We have to make sure that this is something we really need.

CC: @rschlussel @mbasmanova @martint

Reply:
I've changed the concept a bit, see my comment: https://github.com/prestodb/presto/pull/11570#issuecomment-424750443

> Currently i can see only a singe rule that tends to underestimate in some cases. This is FilterStatsRule (in fact FilterStatsCalculator). Also because JoinStatsRule relies on FilterStatsCalculator, the JoinStatsRule may also underestimate.

`FilterStatsRule`, `JoinStatsRule`, `SemiJoinStatsRule`, `SimpleFilterProjectSemiJoinStatsRule`, `SpatialJoinStatsRule` (?). They all rely on `FilterStatsCalculator` which can underestimate.

> From what i understand, most of the underestimation in FilterStatsCalculator are because of the heuristic we use for inequality predicate estimates, as it is very hard to correctly estimate inequality predicates without histograms:

We can also underestimate via correlation. If you have predicates `p1 AND p2 AND p3` depending on join or filter rule we will apply reduction factor for `p1, p2, p3` independently (for join we select so called "driving clause").

> For the nasty cases when we are still underestimating, the user can just disable CBO, and set the BROADCAST distribution manually. Given that our model is not ideal, that shouldn't be very unusual situation.

You only can choose `REPLICATED/BROADCAST` for all joins in a query. The idea is that you would like automatically small tables to be replicated, but large to be partitioned. If you get BROADCAST right you often win big time, but getting it wrong might be a query failure decision.

> I can the value of this change. But from the other side that's another user visible knob with a very limited usability and an area of application. We have to make sure that this is something we really need.

The `broadcast_join_max_table_size` can be left default to some reasonable small value so that only small tables are broadcasted. I would even consider `AUTOMATIC_RESTRICTED` to be set by default as it is relatively safe switch and there is not room for error.

AFAIR some of the distributed databases are very careful with BROADCAST join and probably for a good reason.